🚀 Instate documentation for MLJFlux (#252)
FluxML/MLJFlux.jl: docs/src/index.md

# MLJFlux.jl

> A Julia package integrating deep learning Flux models with MLJ.
> 

Reviewer: ```suggestion
A Julia package integrating deep learning Flux models with [MLJ](https://juliaai.github.io/MLJ.jl/dev/).

```

Although we might need to update it again later :wink:

## Objectives

- Provide a user-friendly and high-level interface to fundamental [Flux](https://fluxml.ai/Flux.jl/stable/) deep learning models while still being extensible by supporting custom models written with Flux

- Make building deep learning models more convenient to users already familiar with the MLJ workflow

- Make it easier to apply machine learning techniques provided by MLJ, including: out-of-sample performance evaluation, hyper-parameter optimization, iteration control, and more, to deep learning models

!!! note "MLJFlux Coverage"

>     MLJFlux support is focused on fundamental and widely used deep learning models; sophisticated architectures or techniques such as online learning, reinforcement learning, and adversarial networks are currently beyond its scope. 

Reviewer: ```suggestion
    MLJFlux support is focused on fundamental and widely used deep learning models.  Sophisticated architectures or techniques such as online learning, reinforcement learning, and adversarial networks are currently beyond its scope. 
```

Also note that MLJFlux is limited to training models only when all training data fits into memory, though it still supports automatic batching of data.

## Installation

```julia
import Pkg
Pkg.activate("my_environment", shared=true)
Pkg.add(["MLJ", "MLJFlux", "Flux"])
```
You only need `Flux` if you need to build a custom architecture or experiment with different optimizers, loss functions and activations.

## Quick Start

> First load and instantiate mode:

Reviewer: ```suggestion
```

```@example
using MLJ, Flux, MLJFlux
import RDatasets

> # 1. Load Data

Reviewer: Have you thought removing the comments like "`# 1. Load Data` and making them ordinary text or headings outside of the fenced code?

Reviewer: Got it! Makes sense.

iris = RDatasets.dataset("datasets", "iris");
y, X = unpack(iris, ==(:Species), colname -> true, rng=123);

# 2. Load and instantiate model
NeuralNetworkClassifier = @load NeuralNetworkClassifier pkg="MLJFlux"
clf = NeuralNetworkClassifier(
    builder=MLJFlux.MLP(; hidden=(5,4), σ=Flux.relu),
    optimiser=Flux.ADAM(0.01),
    batch_size=8,
    epochs=100, 
    acceleration=CUDALibs()         # For GPU support
    )

# 3. Wrap it in a machine in fit
mach = machine(clf, X, y)

> fit!(mach)

Reviewer: We don't need to `fit!` if we're going to evaluate.
```suggestion
```

Author: Oops. Such a rookie mistake. Thanks. 

P.S.: Need to modify the comment as well.

# 4. Evaluate the model
cv=CV(nfolds=5)
evaluate!(mach, resampling=cv, measure=accuracy) 
```
As you can see we were able to use MLJ functionality (i.e., cross validation) with a Flux deep learning model. All arguments provided also have defaults.

> Notice that we were also able to define the neural network in a high-level fashion by only specifying the number of neurons per each hidden layer and the activation function. Meanwhile, `MLJFlux` was able to infer the input and output layer as well as use a suitable default for the loss function and output activation given the classification task. Notice as well that we did not need to implement a training or prediction loop as in `Flux`.

Reviewer: ```suggestion
Notice that we were also able to define the neural network in a high-level fashion by only specifying the number of neurons in each hidden layer and the activation function. Meanwhile, `MLJFlux` was able to infer the input and output layer as well as use a suitable default for the loss function and output activation given the classification task. Notice as well that we did not need to implement a training or prediction loop as in `Flux`.
```

## Basic idea

As in the example above, any MLJFlux model has a `builder` hyperparameter, an object encoding
instructions for creating a neural network given the data that the
model eventually sees (e.g., the number of classes in a classification
problem). While each MLJ model has a simple default builder, users
may need to define custom builders to get optimal results,
and this will require familiarity with the [Flux
API](https://fluxml.ai/Flux.jl/stable/) for defining a neural network
chain.

> In the future MLJFlux may provide a larger assortment of canned
> builders. Pull requests introducing new ones are most welcome.

Reviewer: ```suggestion
```

Author: Somehow I remember removing this before making PR ready for review 🤷.

- **Estimating performance** of your model using a holdout set or other resampling strategy (e.g., cross-validation) as measured by one or more metrics (e.g., loss functions) that may not have been used in training

- **Optimizing hyper-parameters** such as a regularization parameter (e.g., dropout) or a width/height/nchannnels of convolution layer

- **Compose with other models** such as introducing data pre-processing steps (e.g., missing data imputation) into a pipeline. It might make sense to include non-deep learning models in this pipeline. Other kinds of model composition could include blending predictions of a deep learner with some other kind of model (as in “model stacking”). Models composed with MLJ can be also tuned as a single unit.

- **Controlling iteration** by adding an early stopping criterion based on an out-of-sample estimate of the loss, dynamically changing the learning rate (eg, cyclic learning rates), periodically save snapshots of the model, generate live plots of sample weights to judge training progress (as in tensor board)

- **Comparing** your model with a non-deep learning models

> Thus, for model that could be implemented in both `Flux` and `MLJFlux`, one could choose working with `MLJFlux` instead of `Flux` if they are interested in any of the functionality above, while not willing to implement it from scratch and/or when they would prefer working with a more high-level interface equivalent to that of MLJ for their task.

Reviewer: ```suggestion
[`FastAI`](https://fluxml.ai/FastAI.jl/dev/FastAI@dev/doc/docs/introduction.md.html) also provides a high-level interface for interacting with `Flux` models. If you are working exclusively with neural network models, and are less interested with comparison/composition with other kinds of machine learning models, this may be more suitable choice. Otherwise, `MLJFlux` may be better, especially for the users already familiar with the MLJ interface.
```

Author: @ablaom while I have been only browsing and searching in the docs for the last few minutes, my conclusion so far is that `FastAI` does not go too much beyond `providing a high-level interface for interacting with Flux models. `

For instance, I could neither by searching their docs nor quickly browsing the pages find that they support features such as advanced out-of-sample evaluation (e.g., cross validation) or hyperparameter search (e.g., tuning the number of hidden layers/number of neurons per layers). [FluxTraining.jl](https://fluxml.ai/FluxTraining.jl/dev/documents/README.md) (related to FastAI in some way) seems to only support hyperparameter scheduling which is not exactly hyperparameter search (e.g., still can't tune the number of layers with this; can argue it's not really "tuning" whatsoever). That same package does support early stopping though.

Thus, I find the following to not the most accurate fact, in that `FastAI` is not a drop-in replacement for `MLJFlux`:

> If you are working exclusively with neural network models, and are less interested with comparison/composition with other kinds of machine learning models, this (i.e., FastAI) may be more suitable choice.

I'm also not sure why the mention `exclusively with neural network models` because in their README they tend to just speak more generically of deep learning models.

I would recommend one of the following line of actions:
- That we don't mention `FastAI` at all (i.e., what an average package would do; also avoids confusion to users that: (i), only know `MLJFlux` and  (ii), where `MLJFlux` is sufficient for them; if (ii) did not hold then they would have likely been able to find FastAI themselves)
- Go out of our way and maximize sincerity as you intended and mention what we know about `FastAI` (e.g., after the comparison with Flux we add a line like `A comparable project, FastAI/FluxTraining, also provides a high-level interface for interacting with Flux models and supports a subset of features that may overlap with (but not include all of) those supported by MLJFlux`

For the last point, one can also go into more detail in the comparison but I don't find this necessary as both packages are likely to change with time and users may be in general less interested in such comparison. Comparing `MLJFlux` and `Flux` makes lots of sense because it's guaranteed everyone likely knows `Flux` and this helps them understand `MLJFlux` better.

Reviewer: I think it's helpful to make links to other related packages. I also admit, that (out of some laziness) I may have been too generous is stating what FastAI.jl can do. That said, I have had some interactions with the developers and indirectly with the project and have the impression it is a well thought-out design by experts, and I remember a lot of community engagement. If I were working exclusively with images today - and certainly video - I expect I might prefer FastAI to MLJFlux.  

So the second suggestion sounds good, which I agree is more accurate.

Author: Alright. I agree with everything you said. So we will go with replacing the text above with:
> A comparable project, FastAI/FluxTraining, also provides a high-level interface for interacting with Flux models and supports a set of features that may overlap with (but not include all of) those supported by MLJFlux.

Right? Let me know if you would to fine-tune it further.

Reviewer: all good
